File: 2024/python/day4/main.py
import collections  # noqa: F401
import itertools  # noqa: F401
import math  # noqa: F401
import re  # noqa: F401
from dataclasses import dataclass  # noqa: F401
from pathlib import Path

import toolz  # noqa: F401


def part_a(input: str):
    lines = input.splitlines()
    total = 0
    look_for = [tuple("XMAS"), tuple(reversed("XMAS"))]
    # horizontal
    for line in lines:
        for word in toolz.sliding_window(4, line):
            if word in look_for:
                total += 1
    for a, b, c, d in toolz.sliding_window(4, lines):
        vertical = zip(a, b, c, d)
        # diagonal \
        diag1 = zip(a, b[1:], c[2:], d[3:])
        # diagonal /
        diag2 = zip(a[3:], b[2:], c[1:], d)
        for word in itertools.chain(vertical, diag1, diag2):
            if word in look_for:
                total += 1
    return total


def part_b(input: str):
    lines = input.splitlines()
    total = 0
    look_for = [tuple("MAS"), tuple(reversed("MAS"))]
    for a, b, c in toolz.sliding_window(3, lines):
        # diagonal \
        diags1 = zip(a, b[1:], c[2:])
        # diagonal /
        diags2 = zip(a[2:], b[1:], c)
        for diag1, diag2 in zip(diags1, diags2):
            if diag1 in look_for and diag2 in look_for:
                total += 1
    return total


SAMPLE_INPUT = """MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX
"""
SAMPLE_INPUT_B = SAMPLE_INPUT
INPUT = (Path(__file__).parent / "data.txt").read_text()


def test_part_a(capsys):
    assert part_a(SAMPLE_INPUT) == 18
    with capsys.disabled():
        print(f"\nPart A: {part_a(INPUT)}", end=" ")


def test_part_b(capsys):
    assert part_b(SAMPLE_INPUT_B) == 9
    with capsys.disabled():
        print(f"\nPart B: {part_b(INPUT)}", end=" ")

File: 2024/src/bin/day4/main.rs
#![feature(array_windows)]
use std::borrow::Cow;

use advent_2024::{BasicSolution, Solution};
use itertools::izip;

struct Day;

impl BasicSolution for Day {
    const INPUT: &'static str = include_str!("data.txt");
    const SAMPLE_INPUT: &'static str = include_str!("sample.txt");

    type Shared = Vec<&'static [u8]>;
    type Answer = usize;

    const SAMPLE_ANSWER_A: Self::TestAnswer = 18;
    const SAMPLE_ANSWER_B: Self::TestAnswer = 9;

    fn shared(input: &'static str) -> anyhow::Result<Self::Shared> {
        Ok(input.lines().map(str::as_bytes).collect())
    }

    fn part_a(shared: Cow<Self::Shared>) -> anyhow::Result<Self::Answer> {
        const LOOK_FOR: [&[u8; 4]; 2] = [b"XMAS", b"SAMX"];
        Ok(shared
            .array_windows()
            .enumerate()
            .flat_map(|(i, &[a, b, c, d])| {
                let vertical = izip!(a, b, c, d);
                let diag1 = izip!(a, &b[1..], &c[2..], &d[3..]);
                let diag2 = izip!(&a[3..], &b[2..], &c[1..], d);

                let horizontals = if i == 0 {
                    [a, b, c, d]
                } else {
                    [&[], &[], &[], d]
                }
                .into_iter()
                .flat_map(|line| line.array_windows())
                .copied();

                vertical
                    .chain(diag1)
                    .chain(diag2)
                    .map(|(&a, &b, &c, &d)| [a, b, c, d])
                    .chain(horizontals)
            })
            .filter(|word| LOOK_FOR.contains(&word))
            .count())
    }

    fn part_b(shared: Self::Shared) -> anyhow::Result<Self::Answer> {
        const LOOK_FOR: [&[u8; 3]; 2] = [b"MAS", b"SAM"];
        Ok(shared
            .array_windows()
            .flat_map(|&[a, b, c]| izip!(a, &a[2..], &b[1..], c, &c[2..]))
            .filter(|&(&tl, &tr, &m, &bl, &br)| {
                LOOK_FOR.contains(&&[tl, m, br]) && LOOK_FOR.contains(&&[tr, m, bl])
            })
            .count())
    }
}

fn main() -> anyhow::Result<()> {
    Day::main()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a() -> anyhow::Result<()> {
        Day::test_part_a()
    }

    #[test]
    fn b() -> anyhow::Result<()> {
        Day::test_part_b()
    }
}
